### PR TITLE
retrieve the catalog from the broker

### PR DESCRIPTION
**Pull request by qibobo**
[#119505867]

### CODE REVIEW

File: servicebroker_node/config/settings.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+	"port":8080, 
+	
+	"service_id": "autoscaler-guid",
```
Comment (fraenkel) [inline servicebroker_node/config/settings.json]: What is this stuff for?

File: servicebroker_node/lib/api/api.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+module.exports = function() {
+
+  var fs = require('fs');
+  var path = require('path');
+
+  var settings = JSON.parse(
```
Comment (fraenkel) [inline servicebroker_node/lib/api/api.js]: We need to assume the catalog.json and any config are provided via external files.
This needs to be injected rather than hardcoded.
We can create a separate chore if you wish.

File: servicebroker_node/test/ut/api/api_test.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+describe('Unit Test for API', function() {
+
+it('getCatalog() should return catalog info with JSON format', function() {
+  var data = api.getCatalog();
+  data.services[0].id.should.equal(settings.service_id);
+  data.services[0].name.should.equal(settings.service_name);
+  data.services[0].description.should.equal(settings.service_description);
+  data.services[0].plans[0].id.should.equal(settings.plan_id);
+  data.services[0].plans[0].name.should.equal(settings.plan_name);
+  data.services[0].plans[0].description.should.equal(settings.plan_description);
```
Comment (fraenkel) [inline servicebroker_node/test/ut/api/api_test.js]: how do we know that is all that is there?

File: servicebroker_node/test/fat/index_test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+var supertest = require("supertest");
```
Comment (fraenkel) [inline servicebroker_node/test/fat/index_test.js]: This is still a unit test. Just move it over.

File: servicebroker_node/test/fat/index_test.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+var should = require("should");
+var fs = require('fs');
+var path = require('path');
+
+// This agent refers to PORT where program is runninng.
+var settings = JSON.parse(fs.readFileSync( path.join (__dirname, '../../config/settings.json'), 'utf8'));
+var port = process.env.PORT || settings.port;
+var server = supertest.agent("http://localhost:"+port);
```
Comment (fraenkel) [inline servicebroker_node/test/fat/index_test.js]: We should be starting and stopping the server as needed for our testing.
See https://glebbahmutov.com/blog/how-to-correctly-unit-test-express-server/ as an example.

File: servicebroker_node/test/fat/index_test.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+// UNIT test begin
+
+describe("RESTful API test suite",function(){
+
```
Comment (fraenkel) [inline servicebroker_node/test/fat/index_test.js]: What happens with an invalid path?

**Comment by fraenkel** [comment]
You can also just put it in the servicebroker directory. Doesn't have to be servicebroker_node.

File: servicebroker_node/config/settings.json (hunk 1 above)
Comment (cdlliuy) [inline servicebroker_node/config/settings.json]: we extracted the configurable items in catalog here, so that it is easy to edit the key information of catalog with key-value format instead of taking care of the whole catalog JSON format

File: servicebroker_node/lib/api/api.js (hunk 2 above)
Comment (cdlliuy) [inline servicebroker_node/lib/api/api.js]: ok, we did hardcoded the service details and plan details with assumption that we only have 1 plan and 1 service id here.  Maybe back to prior JSON stream approach (save the catalog information in a JSON, and then read it directly) is better.

File: servicebroker_node/config/settings.json (hunk 1 above)
Comment (cdlliuy) [inline servicebroker_node/config/settings.json]: it used to configure the service information and plan information which may different per deployment. i.e. using different service name, service guid and etc.

File: servicebroker_node/test/ut/api/api_test.js (hunk 3 above)
Comment (cdlliuy) [inline servicebroker_node/test/ut/api/api_test.js]: yeah.. it is hardcoded the service details and plan details with assumption that we only have 1 plan and 1 service id here.

File: servicebroker_node/test/fat/index_test.js (hunk 4 above)
Comment (cdlliuy) [inline servicebroker_node/test/fat/index_test.js]: in my understanding,  if we invoke a function directly and verify the input/output , that is unit test.  If we launch the server (here, service broker app) , and invoke the RESTful API to verify the request/response, it is function test. if more apps are launched together, it is integration test.   So, here, we start service broker app , and call /v2/catalog endpoint, it should be a function test.  Is that right?

File: servicebroker_node/test/fat/index_test.js (hunk 6 above)
Comment (cdlliuy) [inline servicebroker_node/test/fat/index_test.js]: do you mean invalid file path in "require" or the invalid http endpoint?  if the file path for "settings" are invalid, the test case should fail.

File: servicebroker_node/lib/api/api.js (hunk 2 above)
Comment (fraenkel) [inline servicebroker_node/lib/api/api.js]: We need to pick one. If it is going to be static, we might has well hardcode it in a file which is read once and always returned. It is a bit of overkill to regenerate the map each time getCatalog is called.

File: servicebroker_node/test/ut/api/api_test.js (hunk 3 above)
Comment (fraenkel) [inline servicebroker_node/test/ut/api/api_test.js]: We should test lengths to guarantee there is no more expected data.

File: servicebroker_node/test/fat/index_test.js (hunk 4 above)
Comment (fraenkel) [inline servicebroker_node/test/fat/index_test.js]: You have units which test everything.
Breaking things down provides no additional value other than unnecessary setup that should be done within the tests anyhow.

Integration tests are exactly that. They test the integration between components which are outside the scope of a single project. It would be similar to where acceptance tests live which requires a live backend. Integration tests create a real setup but not within the target environment,

File: servicebroker_node/config/settings.json (hunk 1 above)
Comment (fraenkel) [inline servicebroker_node/config/settings.json]: I should have asked why we have catalog.json then.

File: servicebroker/lib/api/api.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+var fs = require('fs');
+var path = require('path');
+
+var settings = JSON.parse(
```
Comment (fraenkel) [inline servicebroker/lib/api/api.js]: Why are we loading settings if it isn't used?

File: servicebroker/test/index.html [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
```
Comment (fraenkel) [inline servicebroker/test/index.html]: Is this file really necessary? We can just run mocha.

File: servicebroker/test/mocha.css [hunk 9]
```diff
@@ -0,0 +1,314 @@
+@charset "utf-8";
```
Comment (fraenkel) [inline servicebroker/test/mocha.css]: Same question as above. Also applies to mocha.js.

File: servicebroker/test/ut/api/api_test.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+var should = require('should');
+var API = require(path.join(__dirname, '../../../lib/api/api.js'));
+
+var api = new API();
+
+describe('Unit Test for API', function() {
+
+  it('getCatalog() should return catalog info with JSON format', function() {
+    var data = api.getCatalog();
+    data.should.have.property('services');
```
Comment (fraenkel) [inline servicebroker/test/ut/api/api_test.js]: Shouldn't this match the contents of the catalog.json?

File: servicebroker/test/ut/index_test.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  it("should return catalog json", function(done) {
+    supertest(server)
+      .get("/v2/catalog")
+      .expect("Content-type", /json/)
+      .expect(200) // THis is HTTP response
+      .end(function(err, res) {
+        // HTTP status should be 200
+        res.status.should.equal(200);
+        JSON.stringify(res.body).length.should.equal(JSON.stringify(catalog).length);
+        res.body.services[0].id.should.equal(catalog.services[0].id);
```
Comment (fraenkel) [inline servicebroker/test/ut/index_test.js]: Is this necessary since we already guaranteed it matched the catalog.json.

File: servicebroker/lib/api/api.js (hunk 7 above)
Comment (cdlliuy) [inline servicebroker/lib/api/api.js]: it is used to define a port when start service broker app. It is useless when cf app, but it should be useful if we want to define another port with bosh deploy. Is that right?

File: servicebroker/test/index.html (hunk 8 above)
Comment (cdlliuy) [inline servicebroker/test/index.html]: When we do cmd "mocha init <folder>", then the index.html and others files are automatically added by mocha.

File: servicebroker/test/ut/api/api_test.js (hunk 10 above)
Comment (cdlliuy) [inline servicebroker/test/ut/api/api_test.js]: yes, it should match the content of catalog.json.   
Current unit test purpose is just verify the getCatalog() function return a valid JSON format, and include service definition inside. 
Do you mean we need to parse catalog.json file here, and make sure data=catalog exactly in unit test?

File: servicebroker/lib/api/api.js (hunk 7 above)
Comment (fraenkel) [inline servicebroker/lib/api/api.js]: where? I see it used in index.js not here.

File: servicebroker/test/index.html (hunk 8 above)
Comment (fraenkel) [inline servicebroker/test/index.html]: mocha init is for client side usage not server side.

File: servicebroker/test/ut/index_test.js (hunk 11 above)
Comment (cdlliuy) [inline servicebroker/test/ut/index_test.js]: do you mean we can ignore the test for /v2/catalog since the implementation guaranteed the content should be same?

File: servicebroker/test/ut/api/api_test.js (hunk 10 above)
Comment (fraenkel) [inline servicebroker/test/ut/api/api_test.js]: yes, its more important here than in the index test which can do the same check or validate a catalog jsom document is returned.

File: servicebroker/test/ut/index_test.js (hunk 11 above)
Comment (fraenkel) [inline servicebroker/test/ut/index_test.js]: sorry I missed you only checked length.
you might as well just compare the data which some imply the same length.

File: servicebroker/lib/api/api.js (hunk 7 above)
Comment (cdlliuy) [inline servicebroker/lib/api/api.js]: ah, yes, sorry, this will be removed.

**Comment by cdlliuy** [comment]
Refreshed again based on latest comments. Please help to review

**Comment by fraenkel** [comment]
There should be no basic auth support. It isn't in the story.
We should add another story to cover the security model but it shouldn't be basic auth for the service broker.

File: servicebroker/lib/index.js [hunk 12]
```diff
@@ -14,6 +15,28 @@ var app = express();
 app.use(bodyParser.urlencoded({ extended: true }));
 app.use(bodyParser.json());
 
+var auth = function (req, res, next) {
+  
+  function unauthorized(res) {
+    res.set('WWW-Authenticate', 'Basic realm=Authorization Required');
```
Comment (fraenkel) [inline servicebroker/lib/index.js]: That isn't a realm. A realm must be a quoted string. (https://tools.ietf.org/html/rfc2617)

File: servicebroker/lib/index.js [hunk 13]
```diff
@@ -14,6 +15,28 @@ var app = express();
...
+
+  if (user.name === settings.user && user.pass === settings.password) {
+    return next();
+  } else {
+    return unauthorized(res);
+  };
+  next();
+};
+
+app.use(auth);
```
Comment (fraenkel) [inline servicebroker/lib/index.js]: Don't we want to do this first before we process the body?

File: servicebroker/test/index_test.js [hunk 14]
```diff
@@ -42,8 +46,55 @@ describe("RESTful API test suite", function() {
...
+  });
+
+  it("should return 401 when incorrect auth format provided", function(done) {
+    supertest(server)
+      .get("/v2/catalog")
+      .set("Authorization", "invalidauth"  )
+      .expect(401, done);
+  });
+
+  it("should return 401 when auth info is not encoded with base64", function(done) {
```
Comment (fraenkel) [inline servicebroker/test/index_test.js]: redundant with the above test

File: servicebroker/test/index_test.js [hunk 15]
```diff
@@ -42,8 +46,55 @@ describe("RESTful API test suite", function() {
...
+    supertest(server)
+      .get("/v2/catalog")
+      .set("Authorization", "Basic " + settings.user + ":" + settings.password )
+      .expect(401, done);
+  });
+
+  it("should return 401 when incorrect user/password provided", function(done) {
```
Comment (fraenkel) [inline servicebroker/test/index_test.js]: redundant

File: servicebroker/test/index_test.js [hunk 16]
```diff
@@ -42,8 +46,55 @@ describe("RESTful API test suite", function() {
...
+
+  it("should return 401 when incorrect password provided", function(done) {
+    supertest(server)
+      .get("/v2/catalog")
+      .set("Authorization", "Basic " + new Buffer(settings.user+ ":" + "incorrectpassword").toString('base64') )
+      .expect(401, done);
+
+  });
+
+  it("should return 401 when incorrect basic auth format provided", function(done) {
```
Comment (fraenkel) [inline servicebroker/test/index_test.js]: redundant

File: servicebroker/test/index_test.js [hunk 17]
```diff
@@ -42,8 +46,55 @@ describe("RESTful API test suite", function() {
...
+  });
+
+  it("should return 401 when incorrect user/password provided", function(done) {
+    supertest(server)
+      .get("/v2/catalog")
+      .set("Authorization", "Basic " + new Buffer("incorrectuser:incorrectpassword").toString('base64') )
+      .expect(401, done);
+  });
+
+  it("should return 401 when incorrect password provided", function(done) {
```
Comment (fraenkel) [inline servicebroker/test/index_test.js]: Where is the incorrect user test?

File: servicebroker/test/api/api_test.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+var API = require(path.join(__dirname, '../../lib/api/api.js'));
+
+var catalog = JSON.parse(fs.readFileSync(path.join(__dirname, '../../config/catalog.json'), 'utf8'));
+var api = new API();
+
+describe('Unit Test for API', function() {
+
+  it('getCatalog() should return catalog info with JSON format', function() {
+    var data = api.getCatalog();
+    data.should.have.property('services');
```
Comment (fraenkel) [inline servicebroker/test/api/api_test.js]: This style of testing doesn't make sense.
Why didn't you just do

JSON.stringify(data).should.equal(JSON.stringify(catalog));

Which covers length and content. What you have doesn't guarantee extra data in the catalog.

File: servicebroker/test/index_test.js [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+  it("should return catalog json", function(done) {
+    supertest(server)
+      .get("/v2/catalog")
+      .expect("Content-type", /json/)
+      .expect(200) // THis is HTTP response
+      .end(function(err, res) {
+        // HTTP status should be 200
+        res.status.should.equal(200);
+        JSON.stringify(res.body).length.should.equal(JSON.stringify(catalog).length);
```
Comment (fraenkel) [inline servicebroker/test/index_test.js]: See above.

**Comment by fraenkel** [comment]
I will merge this but the items I identified need to be fixed.

File: servicebroker/lib/index.js (hunk 13 above)
Comment (cdlliuy) [inline servicebroker/lib/index.js]: Yes, using middleware of express, we do the auth first before process the body.  I added console log in code to show the sequence, and result is:
> node lib/index.js

processing the auth first
processing the body later

**Comment by cdlliuy** [comment]
Add enhancement to address the final comment.

**Comment by fraenkel** [comment]
Merged
